gluten-flink/planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecExchange.java

```diff
@@ -0,0 +1,193 @@
+        inputProperty.getRequiredDistribution().getType();
+    switch (distributionType) {
+      case SINGLETON:
+        partitioner = new GlobalPartitioner<>();
+        parallelism = 1;
+        break;
+      case HASH:
+        // TODO Eliminate duplicate keys
+        int[] keys = ((HashDistribution) inputProperty.getRequiredDistribution()).getKeys();
+        InternalTypeInfo<RowData> inputType =
+            (InternalTypeInfo<RowData>) inputTransform.getOutputType();
+        KeySelector keySelector =
+            KeySelectorUtil.getRowDataSelector(
+                planner.getFlinkContext().getClassLoader(), keys, inputType);
+        // --- Begin Gluten-specific code changes ---
+        parallelism = inputTransform.getParallelism();
+        if (inputTransform instanceof OneInputTransformation) {
+          OneInputTransformation oneInputTransform = (OneInputTransformation) inputTransform;
+          if (oneInputTransform.getOperator() instanceof GlutenOperator) {
+            keySelector = new GlutenKeySelector();
+            final ExecEdge inputEdge = getInputEdges().get(0);
+            io.github.zhztheplayer.velox4j.type.RowType glutenInputType =
+                (io.github.zhztheplayer.velox4j.type.RowType)
+                    LogicalTypeConverter.toVLType(inputEdge.getOutputType());
+            io.github.zhztheplayer.velox4j.type.RowType outputType =
+                (io.github.zhztheplayer.velox4j.type.RowType)
+                    LogicalTypeConverter.toVLType(getOutputType());
+            String id = PlanNodeIdGenerator.newId();
+            List<Integer> keyIndexes = Arrays.stream(keys).boxed().collect(Collectors.toList());
+            PartitionFunctionSpec partitionFunctionSpec =
+                new HashPartitionFunctionSpec(glutenInputType, keyIndexes);
+            PlanNode localPartition =
+                new LocalPartitionNode(
+                    id,
+                    List.of(new EmptyNode(outputType)),
+                    "REPARTITION",
+                    false,
+                    partitionFunctionSpec);
+            PlanNode exchange = new StreamPartitionNode(id, localPartition, parallelism);
+            final GlutenSingleInputOperator exchangeKeyGenerator =
+                new GlutenSingleInputOperator(
+                    new StatefulPlanNode(id, exchange),
+                    id,
+                    glutenInputType,
+                    Map.of(id, outputType));
+            inputTransform =
+                ExecNodeUtil.createOneInputTransformation(
+                    inputTransform,
+                    new TransformationMetadata("exchange-hash", "Gluten exchange hash"),
+                    exchangeKeyGenerator,
+                    inputTransform.getOutputType(),
+                    parallelism,
+                    false);
+          }
+        }
+        partitioner =
+            new GlutenKeyGroupStreamPartitioner(keySelector, DEFAULT_LOWER_BOUND_MAX_PARALLELISM);
```

Comment:
This assignment with gluten partitioner is outside the above two if checks. Do you always assume the above two checks pass?

Reply:
Do we need to also keep Flink's original code logic for all such Flink operators to allow fallback?

Reply:
We can not promise fallback everywhere correctly now. But here it does can fallbck.